boa/src/object/internal_methods.rs

```diff
@@ -282,6 +282,52 @@ impl Object {
         })
     }
 
+    /// Essential internal method OwnPropertyKeys
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec](https://tc39.es/ecma262/#table-essential-internal-methods)
+    pub fn own_property_keys(&self) -> Vec<PropertyKey> {
+        let size = self.indexed_properties.len()
+            + self.string_properties.len()
+            + self.symbol_properties.len();
+        let mut prop_keys = Vec::with_capacity(size);
+
+        self.index_properties()
+            .for_each(|(p, _)| prop_keys.push(PropertyKey::from(*p)));
+        self.string_properties()
+            .for_each(|(p, _)| prop_keys.push(PropertyKey::from(p.as_str())));
+        self.symbol_properties()
+            .for_each(|(p, _)| prop_keys.push(PropertyKey::from(p.clone())));
+
+        prop_keys
+    }
+
+    /// The abstract operation ObjectDefineProperties
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-object.defineproperties
+    pub fn define_properties(&mut self, props: &Self) {
```

Comment:
```suggestion
    pub fn define_properties(&mut self, properties: Value) {
```
The properties field can be any value, and `Value::to_object()` is called on it.

Reply:
That also requires a context should i pass one as an argument?

Reply:
> That also requires a context should i pass one as an argument?

yes. we need to pass it, and return Result.

Reply:
Hmm what do you mean it should return a result. I understand that it can return error but what should it return if it succeeds. According to the spec it should return the object but i could not find any way to return it to the value all methods wants a ownership.

Reply:
> Hmm what do you mean it should return a result. I understand that it can return error but what should it return if it succeeds. According to the spec it should return the object but i could not find any way to return it to the value all methods wants a ownership.

We can return `Result<()>`.